# fastCGP
An efficient and exact method to compute gene-level P-values from GWAS single SNP results

(To be modified) One essential question in network-assisted analysis is how to combine SNP-level P-values into a single representative gene-level P-value. A popular strategy is to take the best SNP p-value as gene-level P-value, which is sensitive in capturing high association signals but is also biased to genes with more SNPs assigned. Additional methods accounted for bias of gene size are also suggested, such as [refs]. However, most of them rely on the assumption of independence between test statistics, which is unrealistic because of the well-known LD pattern between SNPs [ref]. Phenotype permutation is known as the gold standard for preservation of the LD pattern, thus is promising for correcting bias of gene size [ref 29]. Yet, for reason of computational intensity and the required access to raw genotyping data, it has never been applied to compute gene-level P-values to our knowledge. An alternative permutation strategy that accounts for the LD pattern while does not require genotyping data is the Circular Genomic Permutation (CGP) strategy [ref]. It permutes SNP-level statistics in a genomic manner to keep similar patterns of correlation in the permutated data as in the original data. 
In this study, we took advantage of the CGP strategy and proposed a novel method fastCGP to compute gene-level P-values. fastCGP corrects the best SNP P-value for gene size while accounts for the correlation between SNPs. Though fastCGP is permutation-based in nature, no sample needs to be generated during computation. More particularly, instead of generating a given number of samples (100,000 for example) to get an empirical estimation, fastCGP used an analytical approach that takes all non-repeating CGP samples into account to obtain the best obtainable p-value. The using of analytical approach brings several advantages. First, the method is exact, hence does not suffer from randomness compared to many simulation-based methods where random numbers are used, such as VEGAS, Truncated Prodcut [refs]â€¦. Second, the computation is efficient. In our study of 2,370,689 SNPs and 24,120 genes, it takes around 0.5 hour on a standard PC (Intel Core i7 3.40GHz CPU, 8GB RAM), which is much faster than many simulation-based competitors [refs]. Third, the resultant P-values are of high precision. According to the formulation of fastCGP, the precision of each gene-level P-value is  , where   is the total number of SNPs analysed in a GWAS. The larger the amount of SNPs is analysed, the higher is the precision. In our case of analysing 2,370,689 SNPs, the precision reaches to 4.2x10-7.

Details of how fastCGP works will be added later.
